### PR TITLE
Each commit is a bugfix 2

### PR DESCRIPTION
**Pull request by kc1212**
The first fix is due to the following, the last fix is a fix of the first fix
* the controller/autoscaler outputs the error in terms of the number of slots/resources, but we actually need the number of instances

The others should be self explanatory

### CODE REVIEW

**Comment by kc1212** [comment]
Ok, this is all for now. With these fixes, and #47, I think we're in pretty good shape. Most of the functionalities are there, I'm able to see tasks getting submitted and the slave would report the task in its heartbeat. I assume the slave is doing its work, but I cannot verify anything because it does not update the progress field in its heartbeat (see #48).

I'm sure there are still bugs. So the next step is to test our system using multiple jobs/slaves, and hopefully flush out more bugs.

File: lib/master/autoscaler.go [hunk 1]
```diff
@@ -24,8 +24,7 @@ func (m *Master) runController() {
 
 	// output is the error in terms of number of resources/slots
 	// we convert it to adjustment to represent the number of instances
-	// NOTE the + 1 may need to be changed if lib.MaxSlotsPerInstance is not 2
-	adjustment := (int(math.Ceil(output)) + 1) / lib.MaxSlotsPerInstance
+	adjustment := int(math.Ceil((output / float64(lib.MaxSlotsPerInstance)) - 1.11e-16))
```
Comment (aaronang) [inline lib/master/autoscaler.go:27]: Can't we just leave out the `1.11e-16` term?

**Comment by aaronang** [comment]
@kc1212 Just had one comment on the `adjustment` computation; somehow GitHub made it outdated instantly.

File: lib/master/autoscaler.go (hunk 1 above)
Comment (kc1212) [reply lib/master/autoscaler.go:27]: It's to prevent the ceiling function changing something like 0.00..01 to 1.

Comment (victor-li) [reply lib/master/autoscaler.go:27]: I also do not like this hacky fix. Can we not solve this in another way?

**Comment by kc1212** [comment]
From my experience, things like this are pretty common. Except maybe you make your epsilon into a constant. 

On 2 November 2016 23:13:27 CET, Victor Li notifications@github.com wrote:

> victor-li commented on this pull request.
> 
> > @@ -24,8 +24,7 @@ func (m *Master) runController() {
> 
>   // output is the error in terms of number of resources/slots
>   // we convert it to adjustment to represent the number of instances
> -  // NOTE the + 1 may need to be changed if lib.MaxSlotsPerInstance is
>   not 2
> -  adjustment := (int(math.Ceil(output)) + 1) / lib.MaxSlotsPerInstance
> -  adjustment := int(math.Ceil((output /
>   float64(lib.MaxSlotsPerInstance)) - 1.11e-16))
> 
> I also do not like this hacky fix. Can we not solve this in another
> way?
> 
> ## 
> 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/aaronang/cong-the-ripper/pull/45

## 

Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Review by aaronang — APPROVED** [review]
If the epsilon thing does the job, I am fine with it.

**Comment by victor-li** [comment]
If it is common, then it is also fine by me.

**Pull request merged**